src/join-9.sol

```diff
@@ -0,0 +1,108 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.5.12;
+
+import "dss/lib.sol";
+
+interface VatLike {
+    function slip(bytes32, address, int256) external;
+}
+
+interface GemLike {
+    function balanceOf(address) external view returns (uint256);
+    function decimals() external view returns (uint8);
+    function transfer(address, uint256) external returns (bool);
+    function transferFrom(address, address, uint256) external returns (bool);
+    function getFeeFor(uint256) external returns (uint256);
+}
+
+// For a token that has a fee (PAXG)
+
+contract GemJoin9 is LibNote {
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external note auth { wards[usr] = 1; }
+    function deny(address usr) external note auth { wards[usr] = 0; }
+    modifier auth {
+        require(wards[msg.sender] == 1, "GemJoin/not-authorized");
+        _;
+    }
+
+    VatLike public vat;   // CDP Engine
+    bytes32 public ilk;   // Collateral Type
+    GemLike public gem;
+    uint256 public dec;
+    uint256 public live;  // Active Flag
+    uint256 public total;
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        wards[msg.sender] = 1;
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        gem = GemLike(gem_);
+        dec = gem.decimals();
+    }
+    function cage() external note auth {
+        live = 0;
+    }
+    function add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x);
+    }
+    function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x);
+    }
+    // Allow dss-proxy-actions to send the gems with only 1 transfer
+    // This should be called via token.transfer() followed by gemJoin.join() atomically or
+    // someone else can steal your tokens
+    function join(address usr) external note returns (uint256) {
+        require(live == 1, "GemJoin/not-live");
+
+        uint256 wad = sub(gem.balanceOf(address(this)), total);
+        require(int256(wad) >= 0, "GemJoin/overflow");
+
+        vat.slip(ilk, usr, int256(wad));
+        total = add(total, wad);
+
+        return wad;
+    }
+    function join(address usr, uint256 wad) external note {
+        require(live == 1, "GemJoin/not-live");
+
+        uint256 sad = sub(wad, gem.getFeeFor(wad));
+        require(int256(sad) >= 0, "GemJoin/overflow");
+
+        vat.slip(ilk, usr, int256(sad));
+        total = add(total, sad);
+
+        uint256 bal = gem.balanceOf(address(this));
+        require(gem.transferFrom(msg.sender, address(this), wad), "GemJoin/failed-transfer");
```

Comment:
I think a better option would be making `join(address usr)` `public` and then having:
```
function join(address usr, uint256 wad) external note {
     require(gem.transferFrom(msg.sender, address(this), wad), "GemJoin/failed-transfer");
     join(usr);
}
```
If I'm not wrong this would make us save two external call and two `SLOAD`s with the caveat that now you will be joining an `amount >= wad` (if by any chance there was balance not recorded in the adapter). However that would be basically the same behaviour than when we are using the other `join` function.

Reply:
Ah yes I like this better. Since the fee is coming out, the assumption is already broken that `wad` = the number of gems you will get, so might as well simplify. This also generalizes the adapter to any token that collects fees, and removes the requirement of `gem.getFeeFor(wad)`. Good call.